### PR TITLE
Implements dpctl.tensor.where

### LINKED ISSUE #1120: Implement dpctl.tensor.where
Function `dpctl.tensor.where` is required by array-API specificaiton ([link](https://data-apis.org/array-api/latest/API_specification/generated/array_api.where.html)).

`dpctl.where(condition, x1, x2) -> array` returns elements chosen from `x1` or `x2` depending on `condition`.

One needs implementation kernel only for `condition` of boolean type, and same data type arrays `x1` and `x2`.

### PR DESCRIPTION
**Pull request by ndgrigorian**
Closes #1120 

This PR implements dpt.where function, which takes a condition array and two arrays of data, x1 and x2. The output is an array with elements from x1 where condition evaluates to true and x2 where condition is false. The output is of type determined by type promotion of x1 and x2.

i.e.,

```
In [4]: condition = dpt.asarray([False, True, True, True])

In [5]: x1 = dpt.asarray([1])

In [6]: x2 = dpt.asarray([2])

In [7]: dpt.where(condition, x1, x2)
Out[7]: usm_ndarray([2, 1, 1, 1])
```

- [X] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [x] Have you checked performance impact of proposed changes?
- [X] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

**Comment by ndgrigorian** [comment]
Significant performance gains from re-implemenation:
```
In [13]: %timeit where(x, x1, x2)
569 µs ± 30.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [14]: %timeit where2(x, x1, x2)
860 µs ± 39.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [15]: %timeit where(x, x1, x2)
524 µs ± 32.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [16]: %timeit where2(x, x1, x2)
851 µs ± 23.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
Where `x` is only 60 elements and `where2` is the original implementation.

For a larger array (on scale of 10^5 elements)

```
In [16]: %timeit wf.where2(x, x1, x2)
1.54 ms ± 71.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [17]: %timeit wf.where(x, x1, x2)
579 µs ± 59 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by oleksandr-pavlyk** [comment]
An issue in type promotion logic:

```python
import dpctl.tensor as dpt

m = dpt.zeros(100, dtype="?")
m[::2] = True; m[1::3] = True; m[2::5] = False

x1 = dpt.ones(100, dtype="i4")
x2 = dpt.zeros(100, dtype="f4")

dpt.where(m, x1, x2)  # raises ValueError
```

The error is `ValueError: Device Intel(R) Graphics [0x9a49] does not provide native support for double-precision floating point type.`.

File: dpctl/tensor/_search_functions.py
Comment (oleksandr-pavlyk) [inline dpctl/tensor/_search_functions.py:79]: The returned array's shape must be consistent with broadcast shape, so I think this check should be moved until after the broadcasting is complete, and it should look like

```python
   if condition.size == 0:
       return dpt.empty(res_shape, dtype=dst_dtype, usm_type=dst_usm_type, sycl_queue=exec_q)
```

**Comment by oleksandr-pavlyk** [comment]
Great to see coverage experience such an upward jump :+1:

File: dpctl/tensor/_search_functions.py
Comment (oleksandr-pavlyk) [inline dpctl/tensor/_search_functions.py:3]:
```suggestion
#  Copyright 2020-2023 Intel Corporation
```

File: dpctl/tensor/libtensor/source/where.cpp
Comment (oleksandr-pavlyk) [inline dpctl/tensor/libtensor/source/where.cpp:102]: Perhaps use a temp variable for `dst_shape[i]` for readability:

```suggestion
        const auto &sh_i = dst_shape[i];
        nelems *= static_cast<size_t>(sh_i);
        shapes_equal = shapes_equal && (x1_shape[i] == sh_i) &&
                       (x2_shape[i] == sh_i) &&
                       (cond_shape[i] == sh_i);
```

File: dpctl/tensor/libtensor/source/where.cpp
Comment (oleksandr-pavlyk) [inline dpctl/tensor/libtensor/source/where.cpp:130]:
```suggestion
        throw py::value_error("Value arrays must have the same data type");
```

File: dpctl/tensor/libtensor/source/where.cpp
Comment (oleksandr-pavlyk) [inline dpctl/tensor/libtensor/source/where.cpp:145]: Sometime down the road it would be good to modularize this check into a utility as well. Perhaps  `dpctl::tensor::utils::can_accomodate(nelems, dst)`. This is best done in a separate PR.

File: dpctl/tensor/_search_functions.py
Comment (oleksandr-pavlyk) [inline dpctl/tensor/_search_functions.py:42]: Please add the docstring.

**Review by oleksandr-pavlyk — CHANGES_REQUESTED** [review]

File: dpctl/tensor/libtensor/source/where.cpp
Comment (ndgrigorian) [reply dpctl/tensor/libtensor/source/where.cpp:145]: I was thinking about this as well during refactoring. I'll make note of it.

**Review by oleksandr-pavlyk — APPROVED** [review]
Looks awesome! Thank you @ndgrigorian

**Review by oleksandr-pavlyk — APPROVED** [review]
Thank you, @ndgrigorian !

**Comment by oleksandr-pavlyk** [comment]
Please wait till github-action CI has finished and merge. This way artifacts would be ready to get published on `dppy/label/dev` right away once the PR is merged.

**Pull request merged**
